include/fiction/algorithms/simulation/sidb/operational_domain.hpp

```diff
@@ -947,46 +1002,57 @@ class operational_domain_impl
      */
     operational_status is_step_point_operational(const step_point& sp) noexcept
     {
-        // if the point has already been sampled, return the stored operational status
-        if (const auto op_value = has_already_been_sampled(sp); op_value.has_value())
+        static std::mutex mutex_to_protect_member_variables;
+
         {
-            return *op_value;
+            const std::lock_guard lock{mutex_to_protect_member_variables};
+            if (const auto op_value = has_already_been_sampled(sp); op_value.has_value())
+            {
+                return *op_value;
+            }
         }
 
-        // fetch the x and y dimension values
         const auto param_point = to_parameter_point(sp);
 
         const auto operational = [this, &param_point]()
         {
+            const std::lock_guard lock(mutex_to_protect_member_variables);
             op_domain.operational_values[param_point] = operational_status::OPERATIONAL;
-
             return operational_status::OPERATIONAL;
         };
 
         const auto non_operational = [this, &param_point]()
         {
+            const std::lock_guard lock(mutex_to_protect_member_variables);
             op_domain.operational_values[param_point] = operational_status::NON_OPERATIONAL;
-
             return operational_status::NON_OPERATIONAL;
         };
 
-        // increment the number of evaluated parameter combinations
-        ++num_evaluated_parameter_combinations;
+        {
+            const std::lock_guard lock(mutex_to_protect_member_variables);
```

Comment:
@marcelwa I think, this mutex is needed and was missing, right?

Reply:
The mutex is actually superfluous here because `num_evaluated_paramter_combinations` is an `std::atomic` type that is inherently thread-safe. I have removed some of the `std::lock_guard`s again that were not needed. Furthermore, I will conduct some testing to see if I can get `phmap`'s built-in thread-safety to work.

Reply:
@marcelwa sorry, I completely missed that! Thank you!

Reply:
As mentioned above, I took the opportunity to really dive into the locking mechanisms of `phmap`'s containers. I think I managed to remove all manual locks in the code and completely rely on submap locking. I therefore adjusted the hash function of `parameter_point` to work with tolerance. This enables a straight-forward map lookup rather than map iteration, which is not thread-safe.